### PR TITLE
feat: Solaris support

### PR DESCRIPTION
**Pull request by larseggert**
Because `libc` on Solaris misses most of the needed symbols, this pulls them in via `bindgen` also for the other BSDs. This has the side effect of removing the dependency on a relatively recent `libc` (= newer than what Gecko requires).

Also some refactoring to the VM CI checks as I was adding Solaris support there.

Drop in coverage is mostly because `llvm-cov` is not available on Solaris.

### CODE REVIEW

**Comment by mxinden** [comment]
> Do we want this as part of 0.2.3?

I don't see why not. Loosening the `libc` requirement is not a breaking change. Am I missing something?

**Comment by larseggert** [comment]
> I don't see why not. Loosening the `libc` requirement is not a breaking change. Am I missing something?

No, you're not missing anything. I thought including this might make landing 0.2.3 in gecko easier, due to the lift of the restriction on `lib`.

Will wait for @martinthomson to give this a review.

File: src/bsd.rs [hunk 1]
```diff
@@ -150,18 +166,19 @@ fn if_name_mtu(idx: u32) -> Result<(String, usize)> {
...
+                    mtu = Some(ifi_mtu);
+                    break;
                 }
             }
-            return Err(default_err());
+            break;
         }
     }
-    Err(default_err())
+    Ok((name.to_string(), mtu))
```
Comment (mxinden) [inline src/bsd.rs:181]: The nesting is difficult for me to parse. What do you think of the following? (Not tested.)

``` rust
    let mtu = IfAddrs::new()?
        .iter()
        .find(|ifa| ifa.addr().sa_family == AF_LINK && ifa.name() == name)
        .and_then(|ifa| ifa.data)
        .and_then(|if_data| usize::try_from(ifa_data.ifi_mtu).ok());
    Ok((name.to_string(), mtu))

```

File: src/bsd.rs [hunk 2]
```diff
@@ -254,20 +287,20 @@ impl From<&RouteMessage> for &[u8] {
...
-    fn from(value: Vec<u8>) -> Self {
+impl From<&[u8]> for rt_msghdr {
+    fn from(value: &[u8]) -> Self {
         debug_assert!(value.len() >= size_of::<Self>());
         unsafe { ptr::read_unaligned(value.as_ptr().cast()) }
     }
 }
 
-fn if_index(remote: IpAddr) -> Result<u16> {
+fn if_index_mtu(remote: IpAddr) -> Result<(u16, Option<usize>)> {
```
Comment (mxinden) [inline src/bsd.rs:297]: How about reducing the nesting in this function a bit? (Not tested.)

``` rust
fn if_index_mtu(remote: IpAddr) -> Result<(u16, Option<usize>)> {
    // Open route socket.
    let mut fd = RouteSocket::new(PF_ROUTE, AF_UNSPEC)?;

    // Send route message.
    let query_seq = RouteSocket::new_seq();
    let query = RouteMessage::new(remote, query_seq)?;
    let query_version = query.version();
    let query_type = query.kind();
    fd.write_all((&query).into())?;

    // Read route messages.
    let pid = unsafe { getpid() };
    loop {
        let mut buf = vec![
            0u8;
            size_of::<rt_msghdr>() +
        // There will never be `RTAX_MAX` sockaddrs attached, but it's a safe upper bound.
         (RTAX_MAX as usize * size_of::<sockaddr_storage>())
        ];
        let len = fd.read(&mut buf[..])?;
        if len < size_of::<rt_msghdr>() {
            return Err(default_err());
        }
        let (reply, mut sa) = buf.split_at(size_of::<rt_msghdr>());
        let reply: rt_msghdr = reply.into();
        if !(reply.rtm_version == query_version
            && reply.rtm_pid == pid
            && reply.rtm_seq == query_seq)
        {
            continue;
        }
        if reply.rtm_type != query_type {
            return Err(default_err());
        }

        // This is a reply to our query.
        // This is the reply we are looking for.
        // Some BSDs let us get the interface index and MTU directly from the reply.
        let mtu: Option<usize> = if reply.rtm_rmx.rmx_mtu != 0 {
            Some(
                reply
                    .rtm_rmx
                    .rmx_mtu
                    .try_into()
                    .map_err(|e: TryFromIntError| unlikely_err(e.to_string()))?,
            )
        } else {
            None
        };
        if reply.rtm_index != 0 {
            // Some BSDs return the interface index directly.
            return Ok((reply.rtm_index, mtu));
        }
        // For others, we need to extract it from the sockaddrs.
        for i in 0..RTAX_MAX {
            if (reply.rtm_addrs & (1 << i)) == 0 {
                continue;
            }
            let saddr = unsafe { ptr::read_unaligned(sa.as_ptr().cast::<sockaddr>()) };
            if saddr.sa_family != AF_LINK {
                (_, sa) = sa.split_at(sockaddr_len(saddr.sa_family)?);
                continue;
            }
            let sdl = unsafe { ptr::read_unaligned(sa.as_ptr().cast::<sockaddr_dl>()) };
            return Ok((sdl.sdl_index, mtu));
        }
        Err(default_err())
    }
}
```

File: src/bsd.rs [hunk 3]
```diff
@@ -170,17 +187,26 @@ union SockaddrStorage {
...
-        unsafe { self.sin.sin_len }
-    }
+fn sockaddr_len(af: AddressFamily) -> Result<usize> {
+    let sa_len = match af {
+        AF_INET => size_of::<sockaddr_in>(),
+        AF_INET6 => size_of::<sockaddr_in6>(),
+        _ => {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Unsupported address family {:af}",
```
Comment (mxinden) [inline src/bsd.rs:197]:
```suggestion
                "Unsupported address family {af:?}",
```

File: src/bsd.rs [hunk 4]
```diff
@@ -285,13 +318,45 @@ fn if_index(remote: IpAddr) -> Result<u16> {
...
+                        if (reply.rtm_addrs & (1 << i)) != 0 {
+                            let saddr =
+                                unsafe { ptr::read_unaligned(sa.as_ptr().cast::<sockaddr>()) };
+                            if saddr.sa_family == AF_LINK {
+                                let sdl = unsafe {
+                                    ptr::read_unaligned(sa.as_ptr().cast::<sockaddr_dl>())
+                                };
+                                return Ok((sdl.sdl_index, mtu));
+                            }
+                            (_, sa) = sa.split_at(sockaddr_len(saddr.sa_family)?);
```
Comment (mxinden) [inline src/bsd.rs:355]: Do I understand correctly that we don't need to `sa.split_at` when `(reply.rtm_addrs & (1 << i)) == 0`?

Comment (larseggert) [reply src/bsd.rs:355]: Yes. There are only as many `sockaddr`s attached to the message as there are set bits in `reply.rtm_addrs` (the set bits signify which response a `sockaddr` for a given position is for. It's a **terrible** interface.

Comment (mxinden) [reply src/bsd.rs:355]: :exploding_head:

**Comment by larseggert** [comment]
@mxinden is there a way to test whether the `libc` version requirement is resolved by this?

**Comment by mxinden** [comment]
@larseggert you should be able to run something along the lines of:

```
cargo update -p libc --precise 0.2.158
# Check `Cargo.lock` whether change to libc version was applied.
cargo check --target {solaris,apple,linux,...}
```

**Comment by larseggert** [comment]
@mxinden thanks, I guess pinning it worked.

**Pull request merged**